swarm/metrics.py

```diff
@@ -0,0 +1,165 @@
+    may vary widely widely in terms of magnitude. So the threshold 0.05 can be used to indicate that the score reaches
+    5% of whatever the score was at the first epoch/batch. For ascending scores, a value greater than one may be
+    helpful. i.e. 20 would indicate improving starting value 20 times over.
+
+    This function also allows you to set a criterion, based on whether it should return the first epoch/batch at which
+    the threshold was crossed, or the first time it was crossed and never thereafter exceeded. (I.e. always remains
+    below that threshold in all tested epochs/batches.)
+    Args:
+        epoch_scores: A list or 1-d Numpy array of floats representing the scores at each epoch/catch.
+        threshold: The nominal value you want the score to first get below or above
+        threshold_type: "ratio" or "nominal". Indicate whether you would like to use the threshold as a nominal value, or a ratio that
+            reflects the change from a reference_iteration. Defaults to "nominal".
+        criterion: "first" or "always". Whether you want to return the epoch at which this threshold is first crossed, or
+            always remains below/above (defaults to first)
+        direction: "above" or "below". Choose whether you want get below or above the threshold (defaults below)
+        reference_iteration: Integer for the epoch/batch at which you want to set your reference to be improved upon.
+            Defaults to 0. Ignored if threshold_type is "nominal".
+
+    Returns: An integer specifying which epoch/batch meets this criterion
+
+    """
+    # First check for suitable inputs. I wish there was a neater way of doing this.
+    valid_threshold_types = {"nominal", "ratio"}
+    valid_criteria = {"first", "always"}
+    valid_directions = {"above", "below"}
+    if threshold_type not in valid_threshold_types:
+        raise ValueError("argument `threshold_type` must be one of %r." % valid_threshold_types)
+    if criterion not in valid_criteria:
+        raise ValueError("argument `criterion` must be in one of %r." % valid_criteria)
+    if direction not in valid_directions:
+        raise ValueError("argument `direcction` must be in one of %r." % valid_directions)
+
+    # Convert to numpy for convenience
+    if type(epoch_scores) == list:
+        epoch_scores = np.array(epoch_scores)
+
+    # Choose what sort of threshold to use, and obtain it
+    if threshold_type == "ratio":
+        reference_score = epoch_scores[reference_iteration]
+        used_threshold = reference_score * threshold
+    elif threshold_type == "nominal":
+        used_threshold = threshold
+
+    # Create boolean mask indicating whether crossed above or below
+    if direction == "above":
+        boolean = epoch_scores >= used_threshold
+    elif direction == "below":
+        boolean = epoch_scores <= used_threshold
```

Comment:
My general philosophy for these kinds of multi-config/multi-option functions is that you simplify the core logic and then have a few functions with these values sort of coded in (like `iteration_threshold_ratio(..)`) This tends to reduce configuration and core logic which is nice. This would make the default nominal and the ratio another function. 

This is a style thing and you'll find examples of both styles out there, this is my opinon

Reply:
Yeah, fair enough. I considered it along the way too. However, I found myself repeating exactly the same bit of code in two or three functions, with just one or two different lines ahead in each one, that I thought against it.  But it is true, that once you've added the config machinery for those different choices, the function does expand quite a bit. If it's purely style, might leave it be, unless I have a sudden change of heart.  The tests are written now for each use case, which I hope makes it viable.

Reply:
To expand on this you'd have `iteration_threshold` that would assume nominal and then `iteration_threshold_ratio` which would convert the args and call `iteration_threshold` so code repetition would be minimal. 
This is a pattern you see a lot in functional languages which is why I program like that a bit

Reply:
Ah true.. that might make it neater. I'll consider.